rust/ranked-prolly-tree/README.md

```diff
@@ -0,0 +1,153 @@
+
+```mermaid
+flowchart TD
+    classDef default fill:#37474F,stroke:#CFD8DC,stroke-width:1px,color:#fff;
+    classDef branch fill:#512DA8;
+    classDef segment fill:#1976D2;
+    classDef leaf fill:#00796B;
+    classDef active fill:#D32F2F;
+    Root["#a380 (max=9)"]:::branch
+    P1["#1f6a (max=2)"]:::segment
+    P2["#4c76 (max=9)"]:::segment
+    C1["1 (R=1)"]:::leaf
+    C2["2 (R=2)"]:::leaf
+    C7["7 (R=1)"]:::leaf
+    C8["8 (R=1)"]:::leaf
+    C9["9 (R=1)"]:::leaf
+    Root --> P1
+    Root --> P2
+    P1 --> C1
+    P1 --> C2
+    P2 --> C7
+    P2 --> C8
+    P2 --> C9 
+```
+
+Nodes are encoded as blocks in the block store, keyed by the hash of the block. **Branches** encode tuples of each child's hash and their boundary/largest key. **Segments** encode tuples of each leaf child's key and value, inlining leaf nodes, such that leaf nodes are not directly addressable from the block store. This strategy allows efficient, lazy lookups, and retrieving additional entries
+
+### Qualities
+
+* Partial sync on demand: A tree can be lazily created given a root hash and possibly remote block store, populating nodes as needed.
+* History independent: Trees containing identical data always arrive at the same shape/root hash, regardless of the order of operations.
+* Non-destructive: As the nodes are represented as blocks in storage, modifying a tree creates new blocks, such that writes do not invalidate previous tree states.
+* Verifiable: As a [merkle tree], each node's hash is comprised of its children hashes, allowing verification of data from the storage layer.
+
+### Ranking
+
+Instead of a boundary function used in prolly trees, we use a ranking function (via [Geometric Search Trees]) to determine ranks of leaf nodes based on their key. At each level in the tree, if a rank is higher than the current rank, a boundary is created. Note the example above, where the leaf with key `2` is rank 2. This creates a boundary, and subsequent leaves have a different parent. Ranks are recursively applied upward, creating new nodes as needed.
+
+Ranking is dependent on some branching factor branching factor (`P` in `Tree<P, _>`), where ranks probabilistically approach each node having approximately `P` children.
+
+> [!NOTE]  
+> In practice, the range for the ranking function is a bit wide, e.g. a ranking factor of 64 could have nodes with 3x as many children. Investigating context-based boundary alternatives, like Dolt's [prolly tree]. 
+
+### Operations
+
+#### Get
+
+Retrieving a value accesses at most *HEIGHT* nodes. For the following example, querying for key `1` from the root, if the key exists it must be within the `#1f6a` node, as it's less than its boundary of `2`. Since `#1f6a` is a segment, all the children are ordered and available, and an entry with key `1` can be quickly found if it exists. 
```

Comment:
```suggestion
Retrieving a value requires traversing number of nodes corresponding to tree height. In the following example, querying for key `1` from the root, if the key exists it must be within the `#1f6a` node, as it's less than its boundary of `2`. Since `#1f6a` is a segment, all the children are ordered and available, and an entry with key `1` can be quickly found if it exists. 
```